compiler/rustc_const_eval/src/const_eval/eval_queries.rs

```diff
@@ -175,25 +175,19 @@ pub(super) fn op_to_const<'tcx>(
         Right(imm) => match *imm {
             _ if imm.layout.is_zst() => ConstValue::ZeroSized,
             Immediate::Scalar(x) => ConstValue::Scalar(x),
-            Immediate::ScalarPair(a, b) => {
-                debug!("ScalarPair(a: {:?}, b: {:?})", a, b);
-                // We know `offset` is relative to the allocation, so we can use `into_parts`.
-                let (data, start) = match a.to_pointer(ecx).unwrap().into_parts() {
-                    (Some(alloc_id), offset) => {
-                        (ecx.tcx.global_alloc(alloc_id).unwrap_memory(), offset.bytes())
-                    }
-                    (None, _offset) => (
-                        ecx.tcx.intern_const_alloc(Allocation::from_bytes_byte_aligned_immutable(
-                            b"" as &[u8],
-                        )),
-                        0,
-                    ),
-                };
-                let len = b.to_machine_usize(ecx).unwrap();
-                let start = start.try_into().unwrap();
-                let len: usize = len.try_into().unwrap();
-                ConstValue::Slice { data, start, end: start + len }
-            }
+            Immediate::ScalarPair(a, b) => ConstValue::ByRef {
+                alloc: ecx.tcx.intern_const_alloc(
```

Comment:
This is silly, we start with an MPlace, then get the immediate, and then intern again?

We should just make sure that `try_as_immediate` is `false` for `ScalarPair`, then this match arm here is unreachable.

Reply:
we... don't always start with an MPlace. I tried this originally.

Reply:
Oh wait no there are more hacks here than I remember. Just saw this comment
```
        // It is guaranteed that any non-slice scalar pair is actually ByRef here.
        // When we come back from raw const eval, we are always by-ref. The only way our op here is
        // by-val is if we are in destructure_mir_constant, i.e., if this is (a field of) something that we
        // "tried to make immediate" before. We wouldn't do that for non-slice scalar pairs or
        // structs containing such.
```
But, still, the ScalarPair arm should then be unreachable, right?

How far away are we from being able to remove destructure_mir_constant entirely, porting all its users to valtrees?

Reply:
I guess this shows that we have an impedence mismatch if `Immediate` can contain `ScalarPair` but we still need to be able to convert arbitrary `OpTy` into `ConstValue`.

FWIW to fix https://github.com/rust-lang/rust/issues/105536 it would probably be enough to change only `try_as_immediate`... then maybe the former `intern_const_alloc` arm here becomes unreachable?

Reply:
> we... don't always start with an MPlace. I tried this originally.

Ah, I think ConstProp will also go through this and it can produce pairs. I hope the new ConstProp will use valtrees instead...